BUG: problems with unit test workflow

While working on #303, noticed that some unit tests have been failing although the unit test workflow continued and reported success.  This is a problem.  Work on this came to overshadow #314 so that PR was split into two parts, #315 and this issue.

- [x] workflow should indicate if pytest had one or more failed tests
- [x] all failed tests should be resolved
- [x] `ModuleNotFoundError: No module named 'suitcase'`


----
Related:

- #303 
- #314
- #315

These tests fail with the `hklpy` environment (databroker v2):

```py
FAILED hkl/tests/test_save_restore_UB.py::test_fourc_orientation_save - assert 0 == 1
FAILED hkl/tests/test_save_restore_UB.py::test_fourc_run_orientation_info - TypeError: Type is not JSON serializable: _ResultTuple
FAILED hkl/tests/test_save_restore_UB.py::test_list_orientation_runs - TypeError: Type is not JSON serializable: _ResultTuple
FAILED hkl/tests/test_save_restore_UB.py::test_no_primary_stream - TypeError: Type is not JSON serializable: _ResultTuple
FAILED hkl/tests/test_save_restore_UB.py::test_missing_energy_key - TypeError: Type is not JSON serializable: _ResultTuple
FAILED hkl/tests/test_save_restore_UB.py::test_restore_orientation - TypeError: Type is not JSON serializable: _ResultTuple
FAILED hkl/tests/test_save_restore_UB.py::test_restore_sample - TypeError: Type is not JSON serializable: _ResultTuple
FAILED hkl/tests/test_util.py::test_issue215 - TypeError: Type is not JSON serializable: _ResultTuple
```

and succeed with `bluesky_2024_1` environment (databroker 1.2):

```py
hkl/tests/test_save_restore_UB.py::test_fourc_orientation_save PASSED                                                                          [ 12%]
hkl/tests/test_save_restore_UB.py::test_fourc_run_orientation_info PASSED                                                                      [ 25%]
hkl/tests/test_save_restore_UB.py::test_list_orientation_runs PASSED                                                                           [ 37%]
hkl/tests/test_save_restore_UB.py::test_no_primary_stream PASSED                                                                               [ 50%]
hkl/tests/test_save_restore_UB.py::test_missing_energy_key PASSED                                                                              [ 62%]
hkl/tests/test_save_restore_UB.py::test_restore_orientation PASSED                                                                             [ 75%]
hkl/tests/test_save_restore_UB.py::test_restore_sample PASSED                                                                                  [ 87%]
hkl/tests/test_util.py::test_issue215 PASSED                                                                                                   [100%]
```

All tests pass locally with databroker v1.2 yet fail with databroker 2.0.  This is a positive situation.  The failures can be attributed to code changes related to databroker v2.  This is understood to be a structural difference.  With databroker v1.2, the documents are from intake while databroker v2 provides tiled Containers.

```py
FAILED hkl/tests/test_util.py::test_issue215 - AssertionError: assert 'fourc' in {}
FAILED hkl/tests/test_util.py::test_RE_documents[None] - AssertionError: assert 'e4cv' in {}
FAILED hkl/tests/test_util.py::test_RE_documents[short] - AssertionError: assert 'e4cv' in {}
FAILED hkl/tests/test_util.py::test_RE_documents[long] - AssertionError: assert 'e4cv' in {}
```

The common point of failure is this code: https://github.com/bluesky/hklpy/blob/dfb7f1861a7462d30d3847f166e66361f90ad8a5/hkl/util.py#L329-L358

--- a/.github/workflows/conda_unit_test.yml
+++ b/.github/workflows/conda_unit_test.yml
@@ -30,14 +30,14 @@ jobs:
         - "3.11"
         databroker-version:
         - "1.2.5"
-        - "2.0.0b31"
+        # - "2.0.0b31"
         include:
           - databroker-version: "1.2.5"
             tiled-version:
             - "0.1.0a96"
-          - databroker-version: "2.0.0b31"
-            tiled-version:
-            - "0.1.0a110"
+          # - databroker-version: "2.0.0b31"
+          #   tiled-version:
+          #   - "0.1.0a110"
 
     steps:
     - uses: actions/checkout@v4

--- a/.github/workflows/conda_unit_test.yml
+++ b/.github/workflows/conda_unit_test.yml
@@ -28,8 +28,8 @@ jobs:
         - "3.10"
         - "3.11"
         databroker-version:
-        - "=1.2"
-        - ">1.3"
+        - "1.2.5"
+        - "2.0.0b31"
         max-parallel: 5
 
     steps:
@@ -49,7 +49,7 @@ jobs:
           pytest
           pytest-cov
           python=${{ matrix.python-version }}
-          databroker ${{ matrix.databroker-version }}
+          databroker=${{ matrix.databroker-version }}
 
     - name: conda environments
       shell: bash -l {0}

--- a/.github/workflows/conda_unit_test.yml
+++ b/.github/workflows/conda_unit_test.yml
@@ -28,36 +28,26 @@ jobs:
         - "3.9"
         - "3.10"
         - "3.11"
-        databroker-version:
-        - "1.2.5"
-        # - "2.0.0b31"
-        include:
-          - databroker-version: "1.2.5"
-            tiled-version:
-            - "0.1.0a96"
-          # - databroker-version: "2.0.0b31"
-          #   tiled-version:
-          #   - "0.1.0a110"
 
     steps:
     - uses: actions/checkout@v4
 
-    - name: Create Python ${{ matrix.python-version }} Databroker ${{ matrix.databroker-version }} environment
+    - name: Create Python ${{ matrix.python-version }} environment
       uses: mamba-org/setup-micromamba@v1
       with:
         cache-environment: true
-        cache-environment-key: env-key-${{ matrix.python-version }}-${{ matrix.databroker-version }}
+        cache-environment-key: env-key-${{ matrix.python-version }}
         condarc: |
           channel-priority: flexible
         environment-file: environment.yml
-        environment-name: hklpy-test-py-${{ matrix.python-version }}-${{ matrix.databroker-version }}
+        environment-name: hklpy-test-py-${{ matrix.python-version }}
         create-args: >-
           coveralls
           pytest
           pytest-cov
           python=${{ matrix.python-version }}
-          databroker=${{ matrix.databroker-version }}
-          tiled=${{ matrix.tiled-version }}
+          databroker=1.2.5
+          tiled=0.1.0a96
 
     - name: conda environments
       shell: bash -l {0}

--- a/.github/workflows/conda_unit_test.yml
+++ b/.github/workflows/conda_unit_test.yml
@@ -31,6 +31,13 @@ jobs:
         databroker-version:
         - "1.2.5"
         - "2.0.0b31"
+        include:
+          - databroker-version: "1.2.5"
+            tiled-version:
+            - "0.1.0a96"
+          - databroker-version: "1.2.5"
+            tiled-version:
+            - "0.1.0a110"
 
     steps:
     - uses: actions/checkout@v4
@@ -50,6 +57,7 @@ jobs:
           pytest-cov
           python=${{ matrix.python-version }}
           databroker=${{ matrix.databroker-version }}
+          tiled=${{ matrix.tiled-version }}
 
     - name: conda environments
       shell: bash -l {0}

--- a/.github/workflows/conda_unit_test.yml
+++ b/.github/workflows/conda_unit_test.yml
@@ -75,4 +75,4 @@ jobs:
       run: |
         coverage run --concurrency=thread --parallel-mode -m pytest -vvv  # --exitfirst .
         coverage combine
-        coverage report--precision 3
+        coverage report

--- a/.github/workflows/conda_unit_test.yml
+++ b/.github/workflows/conda_unit_test.yml
@@ -35,7 +35,7 @@ jobs:
           - databroker-version: "1.2.5"
             tiled-version:
             - "0.1.0a96"
-          - databroker-version: "1.2.5"
+          - databroker-version: "2.0.0b31"
             tiled-version:
             - "0.1.0a110"
 

--- a/.github/workflows/conda_unit_test.yml
+++ b/.github/workflows/conda_unit_test.yml
@@ -27,25 +27,29 @@ jobs:
         - "3.9"
         - "3.10"
         - "3.11"
-      max-parallel: 5
+        databroker-version:
+          - "=1.2"
+          - ">1.3"
+          max-parallel: 5
 
     steps:
     - uses: actions/checkout@v4
-
-    - name: Create Python ${{ matrix.python-version }} environment
+    
+    - name: Create Python ${{ matrix.python-version }} Databroker ${{ matrix.databroker-version }} environment
       uses: mamba-org/setup-micromamba@v1
       with:
         cache-environment: true
-        cache-environment-key: env-key-${{ matrix.python-version }}
+        cache-environment-key: env-key-${{ matrix.python-version }}-${{ matrix.databroker-version }}
         condarc: |
           channel-priority: flexible
         environment-file: environment.yml
-        environment-name: hklpy-test-py-${{ matrix.python-version }}
+        environment-name: hklpy-test-py-${{ matrix.python-version }}-${{ matrix.databroker-version }}
         create-args: >-
           coveralls
           pytest
           pytest-cov
           python=${{ matrix.python-version }}
+          databroker ${{ matrix.databroker-version }}
 
     - name: conda environments
       shell: bash -l {0}

--- a/.github/workflows/conda_unit_test.yml
+++ b/.github/workflows/conda_unit_test.yml
@@ -34,7 +34,7 @@ jobs:
 
     steps:
     - uses: actions/checkout@v4
-    
+
     - name: Create Python ${{ matrix.python-version }} Databroker ${{ matrix.databroker-version }} environment
       uses: mamba-org/setup-micromamba@v1
       with:

--- a/.github/workflows/conda_unit_test.yml
+++ b/.github/workflows/conda_unit_test.yml
@@ -71,6 +71,6 @@ jobs:
     - name: Test with coverage and pytest
       shell: bash -l {0}
       run: |
-        coverage run --concurrency=thread --parallel-mode -m pytest -vvv
+        coverage run --concurrency=thread --parallel-mode -m pytest -vvv  # --exitfirst .
         coverage combine
-        coverage report
+        coverage report--precision 3

--- a/.github/workflows/conda_unit_test.yml
+++ b/.github/workflows/conda_unit_test.yml
@@ -46,8 +46,6 @@ jobs:
           pytest
           pytest-cov
           python=${{ matrix.python-version }}
-          databroker=1.2.5
-          tiled=0.1.0a96
 
     - name: conda environments
       shell: bash -l {0}

--- a/.github/workflows/conda_unit_test.yml
+++ b/.github/workflows/conda_unit_test.yml
@@ -28,9 +28,9 @@ jobs:
         - "3.10"
         - "3.11"
         databroker-version:
-          - "=1.2"
-          - ">1.3"
-          max-parallel: 5
+        - "=1.2"
+        - ">1.3"
+        max-parallel: 5
 
     steps:
     - uses: actions/checkout@v4

--- a/.github/workflows/conda_unit_test.yml
+++ b/.github/workflows/conda_unit_test.yml
@@ -71,6 +71,7 @@ jobs:
     - name: Test with coverage and pytest
       shell: bash -l {0}
       run: |
-        coverage run --concurrency=thread --parallel-mode -m pytest -vvv  # --exitfirst .
+        set -vxeuo pipefail
+        coverage run --concurrency=thread --parallel-mode -m pytest -vvv
         coverage combine
         coverage report

--- a/.github/workflows/conda_unit_test.yml
+++ b/.github/workflows/conda_unit_test.yml
@@ -21,6 +21,7 @@ jobs:
     runs-on: ubuntu-latest
     # needs: lint
     strategy:
+      max-parallel: 5
       matrix:
         python-version:
         - "3.8"
@@ -30,7 +31,6 @@ jobs:
         databroker-version:
         - "1.2.5"
         - "2.0.0b31"
-        max-parallel: 5
 
     steps:
     - uses: actions/checkout@v4

--- a/.github/workflows/publish-docs.yml
+++ b/.github/workflows/publish-docs.yml
@@ -26,7 +26,7 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        python-version: [3.9]
+        python-version: [3.11]
 
     steps:
     - name: Set env vars

--- a/RELEASE_NOTES.rst
+++ b/RELEASE_NOTES.rst
@@ -38,6 +38,7 @@ New Features and/or Enhancements
 Fixes
 -----
 
+* Resolved under-reported problems in CI unit tests.
 * ``util.restore_reflections()`` use renamed motor axes if so defined.
 
 Maintenance

--- a/RELEASE_NOTES.rst
+++ b/RELEASE_NOTES.rst
@@ -53,7 +53,7 @@ Maintenance
 * Documentation ZIP file uploaded as artifact with each build.  Great for review!
 * Expand testing to to Py3.8 - Py3.11.
 * Fix code in ``util.restore_reflections()`` that failed unit tests locally.
-* Pin databroker 1.2 (issue #316).
+* Pin databroker 1.2 (issues #316 & #318).
 * Make it easier to find the SPEC command cross-reference table.
 * Update packaging to latest PyPA recommendations.
 * Validate user input to sample.add_reflection().

--- a/RELEASE_NOTES.rst
+++ b/RELEASE_NOTES.rst
@@ -40,6 +40,7 @@ Fixes
 
 * ``diffract.forward()`` should pick solution consistent with ``diffract.forward_solution_table()``, if it can. Otherwise, fall back to previous iterative method.
 * Resolved under-reported problems in CI unit tests.
+* Raise ValueError in ``util.run_orientation_info()`` for databroker v2+.
 * ``util.restore_reflections()`` use renamed motor axes if so defined.
 
 Maintenance
@@ -52,6 +53,8 @@ Maintenance
 * Documentation ZIP file uploaded as artifact with each build.  Great for review!
 * Expand testing to to Py3.8 - Py3.11.
 * Fix code in ``util.restore_reflections()`` that failed unit tests locally.
+* Pin databroker 1.2 (issue #316).
+* Pin tiled <=0.1.0a96 (issue #316).
 * Make it easier to find the SPEC command cross-reference table.
 * Update packaging to latest PyPA recommendations.
 * Validate user input to sample.add_reflection().

--- a/RELEASE_NOTES.rst
+++ b/RELEASE_NOTES.rst
@@ -54,7 +54,6 @@ Maintenance
 * Expand testing to to Py3.8 - Py3.11.
 * Fix code in ``util.restore_reflections()`` that failed unit tests locally.
 * Pin databroker 1.2 (issue #316).
-* Pin tiled <=0.1.0a96 (issue #316).
 * Make it easier to find the SPEC command cross-reference table.
 * Update packaging to latest PyPA recommendations.
 * Validate user input to sample.add_reflection().

--- a/docs/Makefile
+++ b/docs/Makefile
@@ -3,7 +3,7 @@
 
 # You can set these variables from the command line, and also
 # from the environment for the first two.
-SPHINXOPTS    ?= -vv
+SPHINXOPTS    ?=
 SPHINXBUILD   ?= sphinx-build
 SOURCEDIR     = source
 BUILDDIR      = build

--- a/docs/Makefile
+++ b/docs/Makefile
@@ -3,7 +3,7 @@
 
 # You can set these variables from the command line, and also
 # from the environment for the first two.
-SPHINXOPTS    ?=
+SPHINXOPTS    ?= -vv
 SPHINXBUILD   ?= sphinx-build
 SOURCEDIR     = source
 BUILDDIR      = build

--- a/docs/source/configuration.rst
+++ b/docs/source/configuration.rst
@@ -169,7 +169,67 @@ API
 .. automodule:: hkl.configuration
     :members:
         DiffractometerConfiguration,
-        DCConstraint, DCLattice, DCReflection, DCSample, DCConfiguration,
         _check_key, _check_not_value, _check_range, _check_type, _check_value
     :private-members:
     :undoc-members:
+
+.. autoclass:: hkl.configuration.DCConfiguration
+
+    .. autoattribute:: geometry
+    .. autoattribute:: engine
+    .. autoattribute:: library
+    .. autoattribute:: mode
+    .. autoattribute:: canonical_axes
+    .. autoattribute:: real_axes
+    .. autoattribute:: reciprocal_axes
+    .. autoattribute:: samples
+    .. autoattribute:: samples
+    .. autoattribute:: name
+    .. autoattribute:: datetime
+    .. autoattribute:: wavelength_angstrom
+    .. autoattribute:: energy_keV
+    .. autoattribute:: hklpy_version
+    .. autoattribute:: library_version
+    .. autoattribute:: python_class
+    .. autoattribute:: other
+    .. automethod:: validate
+    .. automethod:: write
+
+.. autoclass:: hkl.configuration.DCConstraint
+
+    .. autoattribute:: low_limit
+    .. autoattribute:: high_limit
+    .. autoattribute:: value
+    .. autoattribute:: fit
+    .. automethod:: validate
+    .. autoproperty:: values
+
+.. autoclass:: hkl.configuration.DCLattice
+
+    .. autoattribute:: a
+    .. autoattribute:: b
+    .. autoattribute:: c
+    .. autoattribute:: alpha
+    .. autoattribute:: beta
+    .. autoattribute:: gamma
+    .. automethod:: validate
+    .. autoproperty:: values
+
+.. autoclass:: hkl.configuration.DCReflection
+
+    .. autoattribute:: reflection
+    .. autoattribute:: position
+    .. autoattribute:: wavelength
+    .. autoattribute:: orientation_reflection
+    .. autoattribute:: flag
+    .. automethod:: validate
+
+.. autoclass:: hkl.configuration.DCSample
+
+    .. autoattribute:: name
+    .. autoattribute:: lattice
+    .. autoattribute:: reflections
+    .. autoattribute:: UB
+    .. autoattribute:: U
+    .. automethod:: validate
+    .. automethod:: write

--- a/env-dev.yml
+++ b/env-dev.yml
@@ -36,6 +36,7 @@ dependencies:
   - spec2nexus
   - sphinx
   - tqdm
+  - trio >=0.22
   - pip:
     - coveralls
     - pytest

--- a/env-dev.yml
+++ b/env-dev.yml
@@ -16,6 +16,7 @@ dependencies:
   - apischema
   - black
   - bluesky
+  - chardet
   - databroker
   - flake8
   - hkl

--- a/environment.yml
+++ b/environment.yml
@@ -29,7 +29,6 @@ dependencies:
   - suitcase-mongo
   - suitcase-msgpack
   - suitcase-utils
-  - tiled
   - tqdm
   - pip:
     - coveralls

--- a/environment.yml
+++ b/environment.yml
@@ -15,6 +15,7 @@ dependencies:
   - python >=3.8,<3.12
   - apischema
   - bluesky
+  - databroker =1.2
   - hkl
   - numpy
   - ophyd

--- a/environment.yml
+++ b/environment.yml
@@ -29,6 +29,7 @@ dependencies:
   - suitcase-mongo
   - suitcase-msgpack
   - suitcase-utils
+  - tiled
   - tqdm
   - pip:
     - coveralls

--- a/environment.yml
+++ b/environment.yml
@@ -25,6 +25,10 @@ dependencies:
   - pygobject
   - pyRestTable
   - spec2nexus
+  - suitcase-jsonl
+  - suitcase-mongo
+  - suitcase-msgpack
+  - suitcase-utils
   - tqdm
   - pip:
     - coveralls

--- a/hkl/calc.py
+++ b/hkl/calc.py
@@ -800,6 +800,11 @@ class CalcRecip(object):
     def __str__(self):
         return repr(self)
 
+    @property
+    def _cfg_reciprocal(self):
+        """Return reciprocal lattice to save as configuration."""
+        return tuple(list(self.sample.reciprocal))
+
 
 class CalcE4CH(CalcRecip):
     """Geometry: E4CH"""

--- a/hkl/configuration.py
+++ b/hkl/configuration.py
@@ -245,7 +245,8 @@ class DCSample:
     (the reciprocal of the crystal) in an orthonormal system.
     """
 
-    U:field(default_factory=lambda: [[]])  # list containing an empty list
+    # TODO: Once py38 is dropped, re-enable the default value setting
+    U: List[List[float]]  # = field(default_factory=list[list[float]])
     """
     Orientation matrix (3 x 3) of the crystal relative to the diffractometer.
     (optional)

--- a/hkl/configuration.py
+++ b/hkl/configuration.py
@@ -38,6 +38,8 @@ import typing
 from dataclasses import asdict
 from dataclasses import dataclass
 from dataclasses import field
+from typing import Dict
+from typing import List
 
 import numpy
 import pyRestTable
@@ -187,13 +189,13 @@ class DCLattice:
 class DCReflection:
     """(internal) Configuration of one orientation reflection."""
 
-    reflection: dict[str, float]
+    reflection: Dict[str, float]
     """
     Reciprocal-space axis positions.  Keys must match in the list of
     ``reciprocal_axes``.
     """
 
-    position: dict[str, float]
+    position: Dict[str, float]
     """
     Real-space axis positions.  Keys must match in the list of
     ``canonical_axes``.
@@ -233,17 +235,18 @@ class DCSample:
     lattice: DCLattice
     """Crystal lattice parameters (angstroms and degrees)"""
 
-    reflections: list[DCReflection]
+    reflections: List[DCReflection]
     """List of orientation reflections."""
 
-    UB: list[list[float]]
+    UB: List[List[float]]
     """
     Orientation matrix (3 x 3).  U is the crystal orientation matrix relative
     to the diffractometer and B is the transition matrix of a non-orthonormal
     (the reciprocal of the crystal) in an orthonormal system.
     """
 
-    U: list[list[float]] = field(default_factory=list[list[float]])
+    # TODO: Once py38 is dropped, re-enable the default value setting
+    U: List[List[float]]  # = field(default_factory=list[list[float]])
     """
     Orientation matrix (3 x 3) of the crystal relative to the diffractometer.
     (optional)
@@ -337,14 +340,14 @@ class DCConfiguration:
     diffractometer to restore.
     """
 
-    canonical_axes: list[str]
+    canonical_axes: List[str]
     """
     List of the diffractometer real-space axis names.  Both the exact spelling
     and order are defined by the back-end computation library.  MUST match
     diffractometer to restore.
     """
 
-    real_axes: list[str]
+    real_axes: List[str]
     """
     User-defined real-space axis names. MUST match diffractometer to restore.
     The length and order of this list must be the same as the
@@ -352,7 +355,7 @@ class DCConfiguration:
     names in this file.
     """
 
-    reciprocal_axes: list[str]
+    reciprocal_axes: List[str]
     """
     List of names of the diffractometer reciprocal-space (pseudo) axes. Both
     the exact spelling and order are defined by the back-end computation
@@ -360,13 +363,13 @@ class DCConfiguration:
     MUST match diffractometer to restore.
     """
 
-    constraints: dict[str, DCConstraint]
+    constraints: Dict[str, DCConstraint]
     """
     Limits to be imposed on the real-space axes for operations and
     computations.  Keys must match in the list of ``canonical_axes``.
     """
 
-    samples: dict[str, DCSample]
+    samples: Dict[str, DCSample]
     """
     Crystalline samples (lattice and orientation reflections).
     The sample name is used as the key in the dictionary.
@@ -410,7 +413,7 @@ class DCConfiguration:
     Name of the Python class that defines this diffractometer. (optional)
     """
 
-    other: dict[str, typing.Any] = field(default_factory=dict)
+    other: Dict[str, typing.Any] = field(default_factory=dict)
     """
     *Any* other content goes into this dictionary (comments, unanticipated
     keys, ...) (optional)

--- a/hkl/configuration.py
+++ b/hkl/configuration.py
@@ -245,8 +245,7 @@ class DCSample:
     (the reciprocal of the crystal) in an orthonormal system.
     """
 
-    # TODO: Once py38 is dropped, re-enable the default value setting
-    U: List[List[float]]  # = field(default_factory=list[list[float]])
+    U:field(default_factory=lambda: [[]])  # list containing an empty list
     """
     Orientation matrix (3 x 3) of the crystal relative to the diffractometer.
     (optional)

--- a/hkl/diffract.py
+++ b/hkl/diffract.py
@@ -177,8 +177,8 @@ class Diffractometer(PseudoPositioner):
     UB = Cpt(AttributeSignal, attr="calc.sample.UB", doc="UB matrix")
     # fmt: off
     reflections = Cpt(
-        ArrayAttributeSignal,
-        attr="calc.sample.reflections",
+        AttributeSignal,
+        attr="_reflections",
         doc="Reflections",
     )
     reflections_details = Cpt(
@@ -448,6 +448,11 @@ class Diffractometer(PseudoPositioner):
     # calculation class, which is probably not a good thing
     # -- it becomes a problem when someone uses these functions
     # outside of move()
+
+    @property
+    def _reflections(self):
+        """Return the list of reflections as a [[float]]"""
+        return [list(r) for r in self.calc.sample.reflections]
 
     @pseudo_position_argument
     def forward(self, pseudo):

--- a/hkl/diffract.py
+++ b/hkl/diffract.py
@@ -168,7 +168,7 @@ class Diffractometer(PseudoPositioner):
     lattice_reciprocal = Cpt(
         # fmt: off
         AttributeSignal,
-        attr="calc.sample.reciprocal",
+        attr="calc._cfg_reciprocal",
         doc="Reciprocal lattice",
         # fmt: on
     )

--- a/hkl/tests/test_util.py
+++ b/hkl/tests/test_util.py
@@ -19,8 +19,6 @@ NO_SUCH_PACKAGE_NAME = "no-such-package"
 
 @pytest.fixture(scope="function")
 def cat():
-    import databroker
-
     yield databroker.temp().v2
 
 

--- a/hkl/tests/test_util.py
+++ b/hkl/tests/test_util.py
@@ -1,7 +1,18 @@
-import pytest
-from packaging import version
+"""Test the hkl.util module."""
 
-import hkl.util
+import json
+import time
+
+import databroker
+import pytest
+from bluesky import plans as bp
+from bluesky.run_engine import RunEngine
+from packaging import version
+from tiled.utils import safe_json_dump
+
+from .. import util
+from .tools import DocsCollector
+from .tools import validate_descriptor_doc_content
 
 NO_SUCH_PACKAGE_NAME = "no-such-package"
 
@@ -78,9 +89,9 @@ def crystal_setup(diffractometer):
 
 
 def test__package_info_states():
-    assert hkl.util._package_info is None
-    hkl.util.get_package_info("hkl")
-    assert hkl.util._package_info is not None
+    assert util._package_info is None
+    util.get_package_info("hkl")
+    assert util._package_info is not None
 
 
 @pytest.mark.parametrize(
@@ -95,7 +106,7 @@ def test__package_info_states():
     ],
 )
 def test_get_package_info(package_name, minimum_version):
-    v = hkl.util.get_package_info(package_name)
+    v = util.get_package_info(package_name)
     if v is None:
         assert package_name in ("hklpy", NO_SUCH_PACKAGE_NAME)
     else:
@@ -120,9 +131,9 @@ def test_get_package_info(package_name, minimum_version):
     # fmt: on
 )
 def test_software_versions_default_list(case):
-    v = hkl.util.software_versions(case)
+    v = util.software_versions(case)
     assert isinstance(v, dict)
-    expected = sorted(hkl.util.DEFAULT_PACKAGE_LIST)
+    expected = sorted(util.DEFAULT_PACKAGE_LIST)
     assert sorted(v.keys()) == expected
 
 
@@ -138,7 +149,7 @@ def test_software_versions_default_list(case):
     ],
 )
 def test_software_versions_items(package_name, minimum_version):
-    v = hkl.util.software_versions([package_name])
+    v = util.software_versions([package_name])
     if package_name in v:
         v_string = v[package_name]
         v_package = version.parse(v_string)
@@ -174,10 +185,11 @@ def test_issue215(cat, RE, fourc):
 
     assert len(cat) == 0
     uids = RE(bp.count([fourc]))
+    time.sleep(1)
     assert len(uids) == 1
     assert len(cat) == 1
 
-    orientation = hkl.util.run_orientation_info(cat[-1])
+    orientation = util.run_orientation_info(cat[-1])
     assert isinstance(orientation, dict)
     assert fourc.name in orientation
     # assert list(orientation.keys()) == []
@@ -191,7 +203,64 @@ def test_issue215(cat, RE, fourc):
     success = False
     try:
         # since it is just a problem of motor names in reflections ...
-        hkl.util.restore_reflections(orient, fourc)
+        util.restore_reflections(orient, fourc)
         success = True
     finally:
         assert success, "Could not restore orientation reflections."
+
+
+@pytest.mark.parametrize("use_refl", [None, "short", "long"])
+def test_RE_documents(use_refl, e4cv):
+    """Issue #316."""
+
+    docs = DocsCollector()  # capture the raw document stream directly
+    assert len(docs.raw) == 0
+
+    # add a reflection?
+    if use_refl == "short":
+        e4cv.calc.sample.add_reflection(1, 2, 3, (40, 0, 0, 60))
+    elif use_refl == "long":
+        # fmt: off
+        e4cv.calc.sample.add_reflection(
+            1, 2, 3,
+            position=e4cv.calc.Position(tth=60, omega=40, chi=0, phi=0),
+        )
+        # fmt: on
+    refs = e4cv.reflections.get()
+    assert isinstance(refs, list)
+    assert len(refs) == len(e4cv.calc.sample.reflections)
+    if len(refs) > 0:
+        assert list(refs[0]) == [1.0, 2.0, 3.0]
+
+    safe = safe_json_dump(refs)
+    assert isinstance(safe, bytes)
+
+    assert isinstance(safe_json_dump(e4cv.reflections_details.get()), bytes)
+    # assert f"{refs=!r}" == "debug"
+
+    RE = RunEngine()
+    cat = databroker.temp().v2
+    RE.subscribe(cat.v1.insert)
+
+    uids = RE(bp.count([e4cv]), docs.receiver)
+    assert len(uids) == 1
+    assert len(docs.raw) == 4
+
+    # Get the configuration from the raw document stream
+    validate_descriptor_doc_content(e4cv.name, docs.raw.get("descriptor"))
+
+    # Get the configuration from the safe_json version of the document stream
+    # difference: np.ndarray --> list
+    # difference: tuple --> list
+    descriptor = docs.safe_json.get("descriptor")
+    assert isinstance(descriptor, bytes)
+    validate_descriptor_doc_content(e4cv.name, json.loads(docs.safe_json.get("descriptor")))
+
+    # Get the configuration from the databroker document stream
+    for key, doc in cat.v1[uids[0]].documents():
+        if key == "descriptor":
+            validate_descriptor_doc_content(e4cv.name, doc)
+
+    orientation = util.run_orientation_info(cat[uids[0]])
+    assert isinstance(orientation, dict)
+    assert e4cv.name in orientation

--- a/hkl/tests/tools.py
+++ b/hkl/tests/tools.py
@@ -7,6 +7,7 @@ Avoids direct imports of __init__.py.
 import logging
 
 import numpy
+from tiled.utils import safe_json_dump
 
 from ..util import new_lattice
 
@@ -14,6 +15,18 @@ logger = logging.getLogger("ophyd_session_test")
 
 TARDIS_TEST_MODE = "lifting_detector_mu"
 TWO_PI = 2 * numpy.pi
+
+
+class DocsCollector:
+    """Collect representative documents from the RE."""
+
+    def __init__(self) -> None:
+        self.raw = {}
+        self.safe_json = {}
+
+    def receiver(self, key, doc):
+        self.raw[key] = doc
+        self.safe_json[key] = safe_json_dump(doc)
 
 
 def new_sample(diffractometer, name, lattice):
@@ -36,3 +49,32 @@ def sample_vibranium(diffractometer):
     a0 = TWO_PI
     cubic = new_lattice(a0)
     return new_sample(diffractometer, "vibranium", lattice=cubic)
+
+
+def validate_descriptor_doc_content(gname, descriptor):
+    """Validate the descriptor document content."""
+    assert isinstance(descriptor, dict)
+
+    config = descriptor.get("configuration")
+    assert isinstance(config, dict)
+
+    diffractometer_config = config.get(gname)
+    assert isinstance(diffractometer_config, dict)
+
+    assert "data" in diffractometer_config
+    data = diffractometer_config.get("data")
+    assert isinstance(data, dict)
+
+    assert isinstance(data[f"{gname}_sample_name"], str)
+    assert isinstance(data[f"{gname}_lattice"], (numpy.ndarray, list))
+    assert isinstance(data[f"{gname}_lattice_reciprocal"], (tuple, list))
+    assert isinstance(data[f"{gname}_U"], (numpy.ndarray, list))
+    assert isinstance(data[f"{gname}_UB"], (numpy.ndarray, list))
+    assert isinstance(data[f"{gname}_reflections_details"], list)
+    assert isinstance(data[f"{gname}__pseudos"], (tuple, list))
+    assert isinstance(data[f"{gname}__reals"], (tuple, list))
+    assert isinstance(data[f"{gname}__constraints"], (numpy.ndarray, list))
+    assert isinstance(data[f"{gname}_ux"], float)
+    assert isinstance(data[f"{gname}_uy"], float)
+    assert isinstance(data[f"{gname}_uz"], float)
+    assert isinstance(data[f"{gname}_orientation_attrs"], list)

--- a/hkl/util.py
+++ b/hkl/util.py
@@ -26,8 +26,6 @@ Also provides `SI_LATTICE_PARAMETER` as defined by the
 .. [#] https://physics.nist.gov/cgi-bin/cuu/Value?asil
 """
 
-from __future__ import print_function
-
 import logging
 import subprocess
 import sys
@@ -341,6 +339,14 @@ def run_orientation_info(run):
     run : from Databroker
         A Bluesky run, from databroker v2, such as ``cat.v2[-1]``.
     """
+    import databroker
+
+    if databroker.__version__ >= "2.0":
+        raise ValueError(
+            f"This function does not work with databroker {databroker.__version__}."
+            "  See https://github.com/bluesky/hklpy/pull/317 for details."
+        )
+
     devices = {}
     try:
         run_conf = run.primary.config

--- a/hkl/util.py
+++ b/hkl/util.py
@@ -344,28 +344,16 @@ def run_orientation_info(run):
     devices = {}
     try:
         run_conf = run.primary.config
-        _aaa = type(run_conf).__name__
         for device in sorted(run_conf):
-            if type(run_conf).__name__ == "Catalog":  # intake catalog
-                conf = run_conf[device].read()
-                if f"{device}_orientation_attrs" in conf:
-                    # fmt:off
-                    devices[device] = {
-                        item[len(device)+1:]: conf[item].to_dict()["data"][0]
-                        for item in conf
-                    }
-                    # fmt:on
-            elif type(run_conf).__name__ == "Container":  # tiled container
-                import yaml
-
-                _bbb = run_conf.read()
-                _bbb = run_conf.item["attributes"]["structure"]
-                conf = run_conf.get(device)
-                print(yaml.dump(conf, indent=2, sort_keys=False))
-            else:
-                raise TypeError(f"Not prepared to handle {type(run_conf)} documents.")
+            conf = run_conf[device].read()
+            if f"{device}_orientation_attrs" in conf:
+                # fmt:off
+                devices[device] = {
+                    item[len(device)+1:]: conf[item].to_dict()["data"][0]
+                    for item in conf
+                }
+                # fmt:on
     except Exception as exc:
-        print(f"{exc=}")
         logger.warning("Could not process run %s, due to %s", run, exc)
     return devices
 

--- a/hkl/util.py
+++ b/hkl/util.py
@@ -344,16 +344,28 @@ def run_orientation_info(run):
     devices = {}
     try:
         run_conf = run.primary.config
+        _aaa = type(run_conf).__name__
         for device in sorted(run_conf):
-            conf = run_conf[device].read()
-            if f"{device}_orientation_attrs" in conf:
-                # fmt:off
-                devices[device] = {
-                    item[len(device)+1:]: conf[item].to_dict()["data"][0]
-                    for item in conf
-                }
-                # fmt:on
+            if type(run_conf).__name__ == "Catalog":  # intake catalog
+                conf = run_conf[device].read()
+                if f"{device}_orientation_attrs" in conf:
+                    # fmt:off
+                    devices[device] = {
+                        item[len(device)+1:]: conf[item].to_dict()["data"][0]
+                        for item in conf
+                    }
+                    # fmt:on
+            elif type(run_conf).__name__ == "Container":  # tiled container
+                import yaml
+
+                _bbb = run_conf.read()
+                _bbb = run_conf.item["attributes"]["structure"]
+                conf = run_conf.get(device)
+                print(yaml.dump(conf, indent=2, sort_keys=False))
+            else:
+                raise TypeError(f"Not prepared to handle {type(run_conf)} documents.")
     except Exception as exc:
+        print(f"{exc=}")
         logger.warning("Could not process run %s, due to %s", run, exc)
     return devices
 

--- a/hkl/util.py
+++ b/hkl/util.py
@@ -365,7 +365,7 @@ def run_orientation_info(run):
                 if f"{device}_orientation_attrs" in conf:
                     # fmt:off
                     devices[device] = {
-                        item[len(device)+1:]: conf[item].to_dict()["data"][0]
+                        item[len(f"{device}_"):]: conf[item].to_dict()["data"][0]
                         for item in conf
                     }
                     # fmt:on

--- a/hkl/util.py
+++ b/hkl/util.py
@@ -353,7 +353,7 @@ def run_orientation_info(run):
                     if f"{device}_orientation_attrs" in conf:
                         # fmt:off
                         devices[device] = {
-                            item[len(device) + 1:]: value
+                            item[len(f"{device}_"):]: value
                             for item, value in conf.items()
                         }
                         # fmt:on

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -49,7 +49,7 @@ classifiers = [
 # pyqt ==5
 dependencies = [
   "apischema",
-  "databroker =1.2",
+  "databroker ==1.2",
   "numpy",
   "ophyd",
   "pint",

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -49,6 +49,7 @@ classifiers = [
 # pyqt ==5
 dependencies = [
   "apischema",
+  "databroker =1.2",
   "numpy",
   "ophyd",
   "pint",
@@ -64,7 +65,7 @@ docs = [
 ]
 tests = [
   "bluesky",
-  "databroker",
+  "tiled <=0.1.0a96",
   "packaging",
 ]
 all = ["hklpy[docs,tests]"]

